### PR TITLE
[API Proposal]: Implement IParsable<TSelf> et cetera on IPEndPoint

### LINKED ISSUE #114405: [API Proposal]: Implement `IParsable<TSelf>` et cetera on `IPEndPoint`
### Background and motivation

`IPEndPoint` has `Parse` methods.  It should implement `IParsable<TSelf>` and the other interfaces that usually go with that.

### API Proposal

Updated by @MihaZupan on 2025-05-20 to add all the methods

```diff
namespace System.Net;

public partial class IPEndPoint : EndPoint
+   , ISpanFormattable, ISpanParsable<IPEndPoint>, IUtf8SpanFormattable, IUtf8SpanParsable<SIPEndPoint>
{
    public static IPEndPoint Parse(ReadOnlySpan<char> s);
    public static IPEndPoint Parse(string s);
    public static bool TryParse(ReadOnlySpan<char> s, [NotNullWhen(true)] out IPEndPoint? result);
    public static bool TryParse(string s, [NotNullWhen(true)] out IPEndPoint? result);

+   public static IPEndPoint Parse(ReadOnlySpan<byte> utf8Text);
+   static IPEndPoint ISpanParsable<IPEndPoint>.Parse(ReadOnlySpan<char> s, IFormatProvider? provider);
+   static IPEndPoint IParsable<IPEndPoint>.Parse(string s, IFormatProvider? provider);
+   static IPEndPoint IUtf8SpanParsable<IPEndPoint>.Parse(ReadOnlySpan<byte> utf8Text, IFormatProvider? provider);

+   public static bool TryParse(ReadOnlySpan<byte> utf8Text, [NotNullWhen(true)] out IPEndPoint? result);
+   static bool ISpanParsable<IPEndPoint>.TryParse(ReadOnlySpan<char> s, IFormatProvider? provider, [NotNullWhen(true)] out IPEndPoint? result);
+   static bool IParsable<IPEndPoint>.TryParse([NotNullWhen(true)] string? s, IFormatProvider? provider, [NotNullWhen(true)] out IPEndPoint? result);
+   static bool IUtf8SpanParsable<IPEndPoint>.TryParse(ReadOnlySpan<byte> utf8Text, IFormatProvider? provider, [NotNullWhen(true)] out IPEndPoint? result);

+   string IFormattable.ToString(string? format, IFormatProvider? formatProvider);
+   public bool TryFormat(Span<char> destination, out int charsWritten);
+   public bool TryFormat(Span<byte> utf8Destination, out int bytesWritten);
+   bool ISpanFormattable.TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format, IFormatProvider? provider);
+   bool IUtf8SpanFormattable.TryFormat(Span<byte> utf8Destination, out int bytesWritten, ReadOnlySpan<char> format, IFormatProvider? provider);
}
```

### API Usage

The type will be usable where generic type constraints require an `IParsable<TSelf>` or other of the added interfaces.

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #114405

### CODE REVIEW

**Comment by ManickaP** [comment]
Hey @pedrobsaila, we are past feature complete for 10.0. So this PR cannot be merged until we branch out and main becomes .NET 11. Please do continue work on the PR, I'll find someone to be responsible for the review. Just letting you know about the state of things right now.

File: src/libraries/System.Net.Primitives/src/System/Net/IPEndPoint.cs [hunk 1]
```diff
@@ -178,5 +179,89 @@ public override int GetHashCode()
...
+            throw new FormatException(SR.bad_endpoint_string);
+        }
+
+        static IPEndPoint ISpanParsable<IPEndPoint>.Parse(ReadOnlySpan<char> s, IFormatProvider? provider) => Parse(s);
+
+        static IPEndPoint IParsable<IPEndPoint>.Parse(string s, IFormatProvider? provider) => Parse(s);
+
+        static IPEndPoint IUtf8SpanParsable<IPEndPoint>.Parse(ReadOnlySpan<byte> utf8Text, IFormatProvider? provider) => Parse(utf8Text);
+
+        public static bool TryParse(ReadOnlySpan<byte> utf8Text, [NotNullWhen(true)] out IPEndPoint? result)
```
Comment (ManickaP) [inline src/libraries/System.Net.Primitives/src/System/Net/IPEndPoint.cs]: Is there a reason for copy-pasting this and having 2 implementations for parsing? 
I'm looking at `IPAddress` and it solves the code duplication with generics:
https://github.com/dotnet/runtime/blob/82f0bd1d38010d82d43855d9ef88da5307652088/src/libraries/System.Net.Primitives/src/System/Net/IPAddressParser.cs#L37-L38

Would it be possible to use the same approach here?

File: src/libraries/System.Net.Primitives/tests/FunctionalTests/IPEndPointTest.cs [hunk 2]
```diff
@@ -260,5 +262,145 @@ public static void Equals_Invoke_ReturnsExpected(IPEndPoint endPoint, object obj
                 Assert.Equal(expected, endPoint.GetHashCode().Equals(obj.GetHashCode()));
             }
         }
+
+        [Fact]
+        public static void ParseByteSpan()
```
Comment (ManickaP) [inline src/libraries/System.Net.Primitives/tests/FunctionalTests/IPEndPointTest.cs]: This needs more test inputs. IPv4, IPv6 full format as well as shortened with `::` etc.
Also there are no negative tests for throwing / returning false and edge cases (null, "").

File: src/libraries/System.Net.Primitives/tests/FunctionalTests/IPEndPointParsing.cs [hunk 3]
```diff
@@ -301,5 +303,337 @@ private void InvalidPortHelper(string addressAndPort)
...
+        }
+
+        [Fact]
+        public static void ParseNullIParsable_ThrowsArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>(() => ParsableHelper<IPEndPoint>.Parse(null, CultureInfo.InvariantCulture));
+        }
+
+        [Theory]
+        [InlineData("10.12.13.-14")]
```
Comment (ManickaP) [inline src/libraries/System.Net.Primitives/tests/FunctionalTests/IPEndPointParsing.cs]: If some of these input data are the same for multiple tests, they could be extracted and shared via `MemberData`.

**Review by ManickaP — APPROVED** [review]
Thank you for being patient. This looks good, ready to merge. I left only a small comment about test, nothing serious.

Thank you!

File: src/libraries/System.Net.Primitives/src/System/Net/IPEndPoint.cs [hunk 4]
```diff
@@ -92,41 +96,65 @@ public static bool TryParse(string s, [NotNullWhen(true)] out IPEndPoint? result
...
+                        return true;
+                    }
+                }
             }
 
             result = null;
             return false;
         }
 
+        public static bool TryParse(ReadOnlySpan<char> s, [NotNullWhen(true)] out IPEndPoint? result) => InternalTryParse(s, out result);
```
Comment (MihaZupan) [inline src/libraries/System.Net.Primitives/src/System/Net/IPEndPoint.cs:160]: Please also include `///` comments on all the new APIs (they seed the api docs)

Comment (ManickaP) [inline src/libraries/System.Net.Primitives/src/System/Net/IPEndPoint.cs:160]:
```suggestion
        /// <summary>Tries to convert the character span to its <see cref="IPEndPoint"/> equivalent.</summary>
        /// <param name="s">A span container the characters representing the <see cref="IPEndPoint"/> to convert.</param>
        /// <param name="result">When this method returns, contains the <see cref="IPEndPoint"/> value equivalent to what is contained in <paramref name="s" /> if the conversion succeeded, or default if the conversion failed. This parameter is passed uninitialized; any value originally supplied in result will be overwritten.</param>
        /// <returns><c>true</c> if <paramref name="s" /> was converted successfully; otherwise, false.</returns>
        public static bool TryParse(ReadOnlySpan<char> s, [NotNullWhen(true)] out IPEndPoint? result) => InternalTryParse(s, out result);
```

**Review by ManickaP — COMMENTED** [review]
-

**Comment by pentp** [comment]
Is it too late to get this into .NET 10? 🙃

**Comment by tannergooding** [comment]
> Is it too late to get this into .NET 10? 🙃

CC. @jeffhandley.

My expectation is yes it is too late. We snapped RC1 already and so this would end up going very hot into RC2, well past the normal cutoff. It isn't as low risk as other changes and was already coming in late, after the cutoff, back in July.

**Comment by ManickaP** [comment]
> Is it too late to get this into .NET 10? 🙃

Yes, it was intentionally held back as the PR went up after feature complete, see https://github.com/dotnet/runtime/pull/118095#issuecomment-3126188551
